### PR TITLE
General code clean-up in transform.py.

### PR DESCRIPTION
**Pull request by owenlittlejohns**
## Description

This PR primarily covers some linting cleanup in `gdal_subsetter/transform.py`. It's not everything, and not meant to be exhaustive, but I had some time today, and wanted to make a start.

The main things covered in this PR:

* Updates to the `checkstackable` class method (now `is_stackable`), including addition of unit tests for it.
* Augmentation of regular expression to find matching `SUBDATASET` output from `gdalinfo` (I'll pop a comment on that to explain it).
* Adoption of `os.path.join()` when concatenating file paths.
* Standardising the use of either double or single quotation marks. PEP8 encourages the use of one or the other, not both. This code had a lot of both. I went with my personal preference of single quotes.
* Adopting f-strings instead of older style "%s" or `str.format()` syntaxes.
* Ensuring the correct amount of whitespace around operators, commas, etc.
* Changing a couple of variable names here and there.
* Some other minor tweaks here and there, probably more than I should have done in one PR.

Note - when HGA v1.2.6 is deployed, the HGA regression tests will need to be updated by merging [this PR](https://github.com/nasa/harmony-regression-tests/pull/42).

## Jira Issue ID

Long-IP, currently doesn't have a ticket.

## Local Test Steps

* Pull this branch locally.
* Build HGA with `bin/build-image`.
* Pull branch for the [accompanying PR for the HGA regression tests](https://github.com/nasa/harmony-regression-tests/pull/42).
* Run those. They should all pass.

## PR Acceptance Checklist
* ~~Jira ticket acceptance criteria met.~~
* [x] `version.txt` and `CHANGE.md` updated if any service code is changed.
* [x] Tests added/updated and passing.
* ~~Documentation updated (if needed).~~

### CODE REVIEW

File: gdal_subsetter/transform.py [hunk 1]
```diff
@@ -1019,107 +1012,92 @@ def get_bbox_lonlat(self, filename):
...
-
-        result_format = True
-        if '.png' in formatlst:
-            result_format = False
-
-        if result_proj and result_gt and result_dtype and result_xsize and result_ysize and result_format:
-            stackable = True
-
-        return stackable
+    def is_stackable(self, files_to_stack: str) -> bool:
```
Comment (owenlittlejohns) [inline gdal_subsetter/transform.py:1015]: This is one of the two main places where I've tweaked things more significantly. The behaviour should be the same as the older version, though (and now we have accompanying unit tests to ensure that functionality persists through future iterations).

Part of me was tempted to extract this into a new module, where it doesn't actually make use of any of the other class attributes or methods, but I thought I'd start with just a quick refactor in place for now.

File: gdal_subsetter/transform.py [hunk 2]
```diff
@@ -1019,107 +1012,92 @@ def get_bbox_lonlat(self, filename):
...
+            ds = gdal.Open(file_to_stack)
+            projections.add(ds.GetProjection())
+            geotransforms.add(str(ds.GetGeoTransform()))
+            dtypes.add(ds.GetRasterBand(1).DataType)
+            xsizes.add(ds.RasterXSize)
+            ysizes.add(ds.RasterYSize)
+            formats.add(os.path.splitext(file_to_stack)[1])
+            ds = None
+
+        same_projection = len(projections) == 1
```
Comment (owenlittlejohns) [inline gdal_subsetter/transform.py:1042]: I could have jammed these conditions directly into the long boolean condition in the return statement, but I reckon this is probably more legible, even though it adds a few lines of code.

File: gdal_subsetter/transform.py [hunk 3]
```diff
@@ -1164,14 +1141,14 @@ def subset2(self, tiffile, outputfile, bbox=None, band=None, shapefile=None):
                 bbox - [left,low,right,upper] in lon/lat coordinates
                 shapefile - a shapefile directory in which multiple files exist
         """
-        process_flags["subset"] = True
+        process_flags['subset'] = True
         # covert to absolute path
         tiffile = os.path.abspath(tiffile)
         outputfile = os.path.abspath(outputfile)
-        #RasterFormat = 'GTiff'
+        # RasterFormat = 'GTiff'
```
Comment (owenlittlejohns) [inline gdal_subsetter/transform.py:1148]: For reference, `pylint` dislikes inline comments that don't have a space after the `#`.

File: gdal_subsetter/transform.py [hunk 4]
```diff
@@ -778,177 +784,161 @@ def stack_multi_file_with_metadata(self, infilelist, outfile):
...
-                (lambda line: re.search(":*" + layer_id+"$", line)),
-                gdalinfo_lines
-            ),
-            None
-        )
+        gdalinfo_lines = self.cmd('gdalinfo', filename)
+
+        layer_line = next((line for line in gdalinfo_lines
+                           if re.search(f'SUBDATASET.*:*{layer_id}$', line)
+                           is not None), None)
```
Comment (owenlittlejohns) [inline gdal_subsetter/transform.py:893]: This is the second of the two more significant changes I made this time around. I noticed that the old version of the HGA regression Sentinel test-case 13 was failing for a silly reason. There is a variable (that probably shouldn't be processed anyway) in the root group called `productBoundingBox`. In the `gdalinfo` output the first line that referred to that variable was part of the metadata:

```
NC_GLOBAL#ogr_layer_name=productBoundingBox
```

However, the thing that should have matched was the subsequent line:

```
SUBDATASET_1_NAME=NETCDF:"/Users/olittlej/Downloads/S1-GUNW-A-R-166-tops-20200313_20200206-014119-34455N_32574N-PP-1749-v2_0_2.nc":productBoundingBox
```

This change guards against that. (It also drops the lambda function in favour of a generator, which I think(?) is considered more "Pythonic")

File: gdal_subsetter/transform.py [hunk 5]
```diff
@@ -1731,17 +1750,17 @@ def _process_geogcs(ds_in, dst):
...
+        dst = Dataset(outfile, mode='w', format='NETCDF4')
         # define global attributes
         dst.title = ''
         dst.institution = 'Alaska Satellite Facility'
         dst.source = ''
         dst.references = ''
         dst.comment = ''
-        dst.history = "{datetime} (UTC)".format(datetime=datetime.utcnow().strftime("%d/%m/%Y %H:%M:%S"))
-        dst.GDAL = "Version {ver}".format(ver=gdal.__version__)
+        dst.history = f'{datetime.utcnow():%d/%m/%Y %H:%M:%S} (UTC)'
```
Comment (owenlittlejohns) [inline gdal_subsetter/transform.py:1760]: I thought this was a pretty cool trick - I didn't previously know that f-strings were smart enough to format datetimes, but I suppose they should handle them as datetimes are part of the Python standard library.

**Review by flamingbear — CHANGES_REQUESTED** [review]
Looks good, had a couple of cleanup items to fix though.

File: gdal_subsetter/transform.py [hunk 6]
```diff
@@ -583,9 +580,9 @@ def get_shapefile(self, subsetshape, dstdir):
 
             if unit and 'Polygon' not in geometry:
                 if "degree" in unit.lower():
```
[the remark this answers is not in the document]
Comment (owenlittlejohns) [reply gdal_subsetter/transform.py:582]: - [ ] Thanks for catching some stagglers. Will update!

File: gdal_subsetter/transform.py [hunk 7]
```diff
@@ -778,177 +784,161 @@ def stack_multi_file_with_metadata(self, infilelist, outfile):
...
-            filter(
-                (lambda line: re.search(":*" + layer_id+"$", line)),
-                gdalinfo_lines
-            ),
-            None
-        )
+        gdalinfo_lines = self.cmd('gdalinfo', filename)
+
+        layer_line = next((line for line in gdalinfo_lines
+                           if re.search(f'SUBDATASET.*:*{layer_id}$', line)
```
[the remark this answers is not in the document]
Comment (owenlittlejohns) [reply gdal_subsetter/transform.py:892]: - [ ] Yeah - I think I prefer to remove that colon all together. I agree that colons can be a bit sneaky in regular expressions. (I _think_ it's performing a match, here, but let's remove any uncertainty.

File: gdal_subsetter/transform.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (owenlittlejohns) [reply gdal_subsetter/transform.py:1015]: - [ ] Good catch - I think we both had half the answer: `List[str]` would probably be the best thing.

Generally - I was a bit half-hearted in adding the types. But I added the ones that were obvious. I think when we understand the code well enough to have all function/method signatures with types, that'll help reading the code further down the road. (Of course, getting that better understanding is 90% of the battle)

File: gdal_subsetter/transform.py [hunk 8]
```diff
@@ -1493,25 +1491,25 @@ def _process_projected(ds_in, dst):
...
+                if attribute_name.find(r'#') > -1:
+                    tmp = attribute_name.split(r'#')
+                    if tmp[0] == 'NC_GLOBAL':
+                        dst.setncattr(tmp[1], attribute_value.rstrip('\n'))
                 else:
-                    if key != '_FillValue' and key != "Conventions":
-                        dst.setncattr(key, glb_attrs[key].rstrip("\n"))
+                    if attribute_name not in ('_FillValue', 'Conventions'):
+                        dst.setncattr(attribute_name,
+                                      attribute_value.rstrip("\n"))
```
[the remark this answers is not in the document]
Comment (owenlittlejohns) [reply gdal_subsetter/transform.py:1502]: - [ ] Update another missed double quotes string.

File: gdal_subsetter/transform.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (flamingbear) [reply gdal_subsetter/transform.py]: I'm pretty sure it's not matching in the example you provided.  the `.*` grabs everything up to layer_id.  `:*` is 0 or more `:`s. and it just gets zero.  ( I was confused so I played in the repl).

**Review by flamingbear — APPROVED** [review]
Looks good. Thanks.

File: gdal_subsetter/transform.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (owenlittlejohns) [reply gdal_subsetter/transform.py]: Fair enough. Probably good to get rid of it, then. Thanks for catching that.

**Pull request merged**